### ISSUE TITLE
Resource Limits for Torque

### ISSUE THREAD

**Issue by dthain** [cooperative-computing-lab/cctools#1857]
@nhazekam please take a break from the makeflow re-write to work on this one:

Mailing list thread:
https://groups.google.com/forum/#!topic/cctools-nd/VfLsGOM_nd4

Please fix up the Makeflow-Torque driver to generate the right options corresponding to the resources specified by a job.  For example, `MEMORY=16 CORES=8` should yield `-l nodes=1:ppn=8;mem=16mb`  Since we know that Torque options can vary, allow the user to override it: if `BATCH_OPTIONS` is set, it should totally overwrite the string.

This should be straightforward, please put out a fix today so it can be tested..

**Comment by stemangiola** [comment]
Hello,

currently if I write 

```
MEMORY=**8000**
CORES=**2**
export BATCH_OPTIONS="-l nodes=1:ppn=2,mem=8gb"
<JOB>

MEMORY=**32000**
CORES=**16**
export BATCH_OPTIONS="-l nodes=1:ppn=2,mem=8gb"
<JOB>
```

I always get the first (or maybe the last, my makeflow file is long) BATCH_OPTIONS executing all the time.

**Comment by nhazekam** [comment]
Hi @stemangiola 

I am glad you have given it a try. Can you try it without the export `BATCH_OPTIONS`?  The Memory and cores should be translated into the command without `BATCH_OPTIONS` now.

Also can you run makeflow with debug options? 
`makeflow <other options> -d all -o makeflow.debug`
and send this to me at nhazekam@nd.edu

**Comment by nhazekam** [comment]
Also, I am making changes now that only specify resources when you are not using BATCH_OPTIONS.

**Comment by stemangiola** [comment]
Hello,

I will update the repository and install it again. I will let you know.

Thanks!

Stefano Mangiola



*Data analyst (MSc, MPhil, PhD candidate) @ Melbourne University *

*Hovens lab - Papenfuss lab*
*Royal Melbourne Hospital - WEHI*
*+61 (0)450135954*

2018-02-20 13:40 GMT+11:00 Nick Hazekamp <notifications@github.com>:

**Comment by stemangiola** [comment]
Hello,

I have the same error as before

Here the log file

`2018/02/21 12:30:55.57 makeflow[2837] makeflow: COLON
2018/02/21 12:30:55.57 makeflow[2837] makeflow: LITERAL: benchmark/pure-N52_ARNA_input.RData
2018/02/21 12:30:55.57 makeflow[2837] makeflow: NEWLINE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: COMMAND:
2018/02/21 12:30:55.58 makeflow[2837] makeflow: COMMAND_MOD_END:
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: mkdir
2018/02/21 12:30:55.58 makeflow[2837] makeflow: SPACE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: -p
2018/02/21 12:30:55.58 makeflow[2837] makeflow: SPACE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: benchmark/pure-N52_ARNA_armet_0.01_5_10;
2018/02/21 12:30:55.58 makeflow[2837] makeflow: SPACE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: Rscript
2018/02/21 12:30:55.58 makeflow[2837] makeflow: SPACE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: test_armet.R
2018/02/21 12:30:55.58 makeflow[2837] makeflow: SPACE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: benchmark/pure-N52_ARNA_input.RData
2018/02/21 12:30:55.58 makeflow[2837] makeflow: SPACE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: benchmark/pure-N52_ARNA_armet_0.01_5_10/results.csv
2018/02/21 12:30:55.58 makeflow[2837] makeflow: NEWLINE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: node command=mkdir -p benchmark/pure-N52_ARNA_armet_0.01_5_10; Rscript test_armet.R benchmark/pure-N52_ARNA_input.RData benchmark/pure-N52_ARNA_armet_0.01_5_10/results.csv
2018/02/21 12:30:55.58 makeflow[2837] makeflow: VARIABLE: =
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: MEMORY
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: 32000
2018/02/21 12:30:55.58 makeflow[2837] makeflow: dag appending to variable name=MEMORY, value=32000
2018/02/21 12:30:55.58 makeflow[2837] makeflow: VARIABLE: =
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: CORES
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: 32
2018/02/21 12:30:55.58 makeflow[2837] makeflow: dag appending to variable name=CORES, value=32
2018/02/21 12:30:55.58 makeflow[2837] makeflow: FILES:
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: benchmark/pure-N52_ARNA_armet_0.01_5_20/results.csv
2018/02/21 12:30:55.58 makeflow[2837] makeflow: COLON
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: benchmark/pure-N52_ARNA_input.RData
2018/02/21 12:30:55.58 makeflow[2837] makeflow: NEWLINE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: COMMAND:
2018/02/21 12:30:55.58 makeflow[2837] makeflow: COMMAND_MOD_END:
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: mkdir
2018/02/21 12:30:55.58 makeflow[2837] makeflow: SPACE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: -p
2018/02/21 12:30:55.58 makeflow[2837] makeflow: SPACE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: benchmark/pure-N52_ARNA_armet_0.01_5_20;
2018/02/21 12:30:55.58 makeflow[2837] makeflow: SPACE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: Rscript
2018/02/21 12:30:55.58 makeflow[2837] makeflow: SPACE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: test_armet.R
2018/02/21 12:30:55.58 makeflow[2837] makeflow: SPACE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: benchmark/pure-N52_ARNA_input.RData
2018/02/21 12:30:55.58 makeflow[2837] makeflow: SPACE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: LITERAL: benchmark/pure-N52_ARNA_armet_0.01_5_20/results.csv
2018/02/21 12:30:55.58 makeflow[2837] makeflow: NEWLINE
2018/02/21 12:30:55.58 makeflow[2837] makeflow: node command=mkdir -p benchmark/pure-N52_ARNA_armet_0.01_5_20; Rs`

**Comment by nhazekam** [comment]
Do you have the logfile as a file and not a screenshot?

**Comment by nhazekam** [comment]
Also, something to note about Makeflow resource specifications that may be causing issues for you. Categories are used to specify resources for a group of nodes. When you specify `MEMORY` and `CORES` they are being attributed to the default category. This is also true of `BATCH_OPTIONS`. As your specification is currently written I believe your values for all three are overwritten each time you re-specify them and when Makeflow is run it uses the last specification. If you want to test if this is the problem add `CATEGORY='NAME'` and specify a unique name for each group. This should not rely on a new installation, just adding CATEGORY lines.

**Comment by stemangiola** [comment]
Sorry I should have been more specific,

I got the same error as 

```
MEMORY=16
CORES=8
benchmark/Fig3_LM22+_armet_1_0.5_2/results.csv: benchmark/Fig3_LM22+_input.RData
        ${MKDIR_P} benchmark/Fig3_LM22+_armet_1_0.5_2; Rscript test_armet.R benchmark/Fig3_LM22+_input.RData benchmark/Fig3_LM22+_armet_1_0.5_2/results.csv

Torque returns this error

--------------------------------QSUB SUBMIT ERROR--------------------------------
Resource specifications incomplete or not found
Please specify number of nodes, number of ppn, and memory required

Correct resource syntax:

        qsub -l nodes=1:ppn=1,mem=1gb
OR      #PBS -l nodes=1:ppn=1,mem=1gb
---------------------------------------------------------------------------------
```


[makeflow.zip](https://github.com/cooperative-computing-lab/cctools/files/1745948/makeflow.zip)

**Comment by nhazekam** [comment]
Thank you for the info. I appreciate the debug log. Can you specify what your command to run this makeflow is?

**Comment by stemangiola** [comment]
Hello,

here my command and the error

` makeflow -T torque -N sr --max-remote 1 -d all -o makeflow.debug
parsing ./Makeflow...
local resources: 4 cores, 15887 MB memory, 423837048 MB disk
max running remote jobs: 1
max running local jobs: 4
checking ./Makeflow for consistency...
./Makeflow has 5083 rules.
recovering from log file ./Makeflow.makeflowlog...
starting workflow....
submitting job: Rscript /home/mangiola.s/PhD/deconvolution/test_ARMET/myTest_tcga_cc_a_TME/create_input.R benchmark/name1-norm_ARNA_input.RData

--------------------------------QSUB SUBMIT ERROR--------------------------------
Resource specifications incomplete or not found
Please specify number of nodes, number of ppn, and memory required

Correct resource syntax:

        qsub -l nodes=1:ppn=1,mem=1gb
OR      #PBS -l nodes=1:ppn=1,mem=1gb

For further information, please refer to the Research Computing Page on Catalyst:
http://catalyst.wehi.edu.au/research/computationalbiology/pages/Research-computing.aspx

----------------------------------------------------------------------------------

qsub: Your job has been administratively rejected by the queueing system.
qsub: There may be a more detailed explanation prior to this notice.
2018/02/22 10:04:15.50 makeflow[15676] notice: job submission failed: no output from torque
couldn't submit batch job, still trying...
^Creceived signal 2 (Interrupt), cleaning up remote jobs and files...
got abort signal...
aborting remote job 0
workflow was aborted.
`
Here, my Makefile

` head -40 Makeflow
MKDIR_P=mkdir -p
METH_DIR=${HOME}/PhD/deconvolution/test_ARMET
MEMORY=8000
CORES=2
benchmark/Fig3_LM22_cibersort/results.csv: benchmark/Fig3_LM22_input.RData
        ${MKDIR_P} benchmark/Fig3_LM22_cibersort; Rscript test_cibersort.R benchmark/Fig3_LM22_input.RData benchmark/Fig3_LM22_cibersort/results.csv
MEMORY=8000
CORES=2
benchmark/Fig3_LM22_llsr/results.csv: benchmark/Fig3_LM22_input.RData
        ${MKDIR_P} benchmark/Fig3_LM22_llsr; Rscript test_llsr.R benchmark/Fig3_LM22_input.RData benchmark/Fig3_LM22_llsr/results.csv
MEMORY=8000
CORES=2
benchmark/Fig3_LM22redo_cibersort/results.csv: benchmark/Fig3_LM22redo_input.RData
        ${MKDIR_P} benchmark/Fig3_LM22redo_cibersort; Rscript test_cibersort.R benchmark/Fig3_LM22redo_input.RData benchmark/Fig3_LM22redo_cibersort/results.csv
MEMORY=16000
CORES=8
benchmark/Fig3_LM22redo_armet_1_0.1_1/results.csv: benchmark/Fig3_LM22redo_input.RData
        ${MKDIR_P} benchmark/Fig3_LM22redo_armet_1_0.1_1; Rscript test_armet.R benchmark/Fig3_LM22redo_input.RData benchmark/Fig3_LM22redo_armet_1_0.1_1/results.csv
MEMORY=16000
CORES=8
benchmark/Fig3_LM22redo_armet_1_0.1_2/results.csv: benchmark/Fig3_LM22redo_input.RData
        ${MKDIR_P} benchmark/Fig3_LM22redo_armet_1_0.1_2; Rscript test_armet.R benchmark/Fig3_LM22redo_input.RData benchmark/Fig3_LM22redo_armet_1_0.1_2/results.csv
MEMORY=16000
CORES=8
benchmark/Fig3_LM22redo_armet_1_0.1_5/results.csv: benchmark/Fig3_LM22redo_input.RData
        ${MKDIR_P} benchmark/Fig3_LM22redo_armet_1_0.1_5; Rscript test_armet.R benchmark/Fig3_LM22redo_input.RData benchmark/Fig3_LM22redo_armet_1_0.1_5/results.csv
MEMORY=16000
CORES=8
benchmark/Fig3_LM22redo_armet_1_0.1_10/results.csv: benchmark/Fig3_LM22redo_input.RData
        ${MKDIR_P} benchmark/Fig3_LM22redo_armet_1_0.1_10; Rscript test_armet.R benchmark/Fig3_LM22redo_input.RData benchmark/Fig3_LM22redo_armet_1_0.1_10/results.csv
MEMORY=16000
CORES=8
benchmark/Fig3_LM22redo_armet_1_0.1_20/results.csv: benchmark/Fig3_LM22redo_input.RData
        ${MKDIR_P} benchmark/Fig3_LM22redo_armet_1_0.1_20; Rscript test_armet.R benchmark/Fig3_LM22redo_input.RData benchmark/Fig3_LM22redo_armet_1_0.1_20/results.csv
MEMORY=16000
CORES=8
benchmark/Fig3_LM22redo_armet_1_0.1_50/results.csv: benchmark/Fig3_LM22redo_input.RData
        ${MKDIR_P} benchmark/Fig3_LM22redo_armet_1_0.1_50; Rscript test_armet.R benchmark/Fig3_LM22redo_input.RData benchmark/Fig3_LM22redo_armet_1_0.1_50/results.csv
MEMORY=16000
CORES=8
`

**Comment by nhazekam** [comment]
Thanks! I am sorry to continue to ask for stuff, but could you also let me know what the result of 

`echo $BATCH_OPTIONS` is? 

From looking at your commandline and Makeflow it seems that BATCH_OPTIONS is getting pulled from somewhere else. I also looked into our source and in Makeflow we pull `BATCH_OPTIONS` from the environment.

**Comment by stemangiola** [comment]
Hello,

such variable is empty. IT tech says to the question is such variable is set somewhere "I don’t think so. Torque (formerly pbs) seems to use variables that start with PBS_. "

Just to clarify:

-B option work beautifully, however I would like to se a different BATCH_OPTIONS for every command in the Makefile, because I have commands that requite vastly different resources, at the moment I am asking a lot of resources for all commands, which is really not ideal.

Thanks.

**Comment by nhazekam** [comment]
If you are able can, run Makeflow using BATCH_OPTIONS and also CATEGORY. 

```
CATEGORY='CAT_1'
MEMORY=8000
CORES=2
export BATCH_OPTIONS="-l nodes=1:ppn=2,mem=8gb"
<JOB>
```

I agree that the normal options are not specifying what you are looking for at the moment. Unfortunately We don't have access to a Torque cluster at the moment.

Are you able to install your own software? If so you should be able to install CCTools and test it out.

**Comment by stemangiola** [comment]
Do I have to specify a different CATEGORY for every job?
For example:

```
CATEGORY='CAT_1'
MEMORY=8000
CORES=2
export BATCH_OPTIONS="-l nodes=1:ppn=2,mem=8gb"
<JOB1>
CATEGORY='CAT_2'
MEMORY=8000
CORES=16
export BATCH_OPTIONS="-l nodes=1:ppn=2,mem=8gb"
<JOB2>
CATEGORY='CAT_3'
MEMORY=8000
CORES=16
export BATCH_OPTIONS="-l nodes=1:ppn=2,mem=8gb"
<JOB3>
```

**Comment by nhazekam** [comment]
You should only have to once for each set of similar tasks. This should also be true for the other resources.

You can look at this for an example:
http://ccl.cse.nd.edu/software/manuals/makeflow.html#rescat

**Comment by nhazekam** [comment]
@stemangiola can you install CCTools in you userspace to test this most recent patch?

You can check out my repo from git or pull the code from #1858 as a zip:
https://github.com/nhazekam/cctools/archive/torque_resources.zip

CCTools does not require permissions to install in you user space.

**Comment by stemangiola** [comment]
I will, I need to finish a big batch of analyses first.

Just a feedback on your comment "_You should only have to once for each set of similar tasks. This should also be true for the other resources_.". Unfortunately didn't fix the problem. There are two cases:

1) I use BATCH_OPTION and no matter what I do, execute just the first exported one
2) I do NOT use BATCH_OPTION, and it does not work, with torque giving error

Bw.

**Comment by stemangiola** [comment]
@nhazekam : "Unfortunately We don't have access to a Torque cluster at the moment"

As far as I know torque can be installed on just one node, for testing purposes (https://www.discngine.com/blog/2014/6/27/install-torque-on-a-single-node-centos-64)

But I imagine can be harder than it seems

**Comment by nhazekam** [comment]
I think this was addressed in #1858. Please respond to this issue if that is not the case.
